### ISSUE TITLE
Kong Balancer timing out for a few minutes after starting

### ISSUE THREAD

**Issue by dliberman** [Kong/kong#4507]
We're facing a problem with Kong 1.1.0 **dbless**, using declarative config. If in our config there's an upstream created, Kong will timeout for a few seconds/minutes after restarting, until it apparently stabilizes.

I ran some investigation and it seems that when we're using a database (in our case, postgres), after starting it will not try to create the balancers for all upstreams upon initialization. But on dbless mode, it will create the balancers after parsing the declarative config. [Here](https://github.com/Kong/kong/blob/bcba8cb22769e741605429520695f2d37f459961/kong/runloop/handler.lua#L552), basically:

```lua
      ngx.timer.at(0, function()
        balancer.init()
      end)
```

The problem is, using more than 1 worker there seems to be some problem that makes some of the workers waiting for long times until the balancers get created. Still couldn't figure out what's the concurrency issue, but I've added some logs to investigate (they're on the attached patch, I've put them over [tag 1.1.0](https://github.com/Kong/kong/tree/1.1.0)).
So some thread or worker marks that ```creating[upstream.id]``` as true and other workers keep waiting for a long time on that ```wait(upstream.id)``` function (talking about [this code](https://github.com/Kong/kong/blob/bcba8cb22769e741605429520695f2d37f459961/kong/runloop/balancer.lua#L367)), not sure why the thread actually creating the balancer is holding this and not releasing the creating state.

We can see the difference on the logs here (using ```tail - f error.log | grep -E "(balancer\.lua|BALANCER_DBLESS_DEBUG)"```):

##### 1 worker:
```
2019/04/12 14:06:21 [debug] 6827#0: *1 [lua] handler.lua:551: before(): BALANCER_DBLESS_DEBUG - calling balancer.init()
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:86: fetching upstream: 3a3f34b0-d4f8-424c-aba7-70822f46c8f0
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:742: init(): BALANCER_DBLESS_DEBUG - before creating balancer
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:363: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 1
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:368: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 2
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:377: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 3
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:385: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 4
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:390: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 5
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:112: fetching targets for upstream: 3a3f34b0-d4f8-424c-aba7-70822f46c8f0
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:401: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 6
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:409: create_balancer(): BALANCER_DBLESS_DEBUG - Marking creating with nil
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:744: init(): BALANCER_DBLESS_DEBUG - after creating balancer
2019/04/12 14:06:21 [debug] 6827#0: *2 [lua] balancer.lua:752: init(): initialized 1 balancer(s), 0 error(s)
```

##### 2 workers:
```
2019/04/12 14:08:30 [debug] 7023#0: *2 [lua] handler.lua:551: before(): BALANCER_DBLESS_DEBUG - calling balancer.init()
2019/04/12 14:08:30 [debug] 7022#0: *1 [lua] handler.lua:551: before(): BALANCER_DBLESS_DEBUG - calling balancer.init()
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:86: fetching upstream: c81ff421-7a46-4d2d-9ce5-f560d6a4c455
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:742: init(): BALANCER_DBLESS_DEBUG - before creating balancer
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:363: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 1
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:368: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 2
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:377: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 3
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:385: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 4
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:390: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 5
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:112: fetching targets for upstream: c81ff421-7a46-4d2d-9ce5-f560d6a4c455
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:401: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 6
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:742: init(): BALANCER_DBLESS_DEBUG - before creating balancer
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:363: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 1
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:368: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 2
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:377: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 3
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:385: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 4
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:390: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 5
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:401: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 6
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:409: create_balancer(): BALANCER_DBLESS_DEBUG - Marking creating with nil
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:744: init(): BALANCER_DBLESS_DEBUG - after creating balancer
2019/04/12 14:08:31 [debug] 7023#0: *3 [lua] balancer.lua:752: init(): initialized 1 balancer(s), 0 error(s)
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:363: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 1
2019/04/12 14:08:31 [debug] 7022#0: *7 [lua] balancer.lua:368: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 2
2019/04/12 14:08:31 [debug] 7023#0: *35 [lua] balancer.lua:363: create_balancer(): BALANCER_DBLESS_DEBUG - creating balancer - step 1
2019/04/12 14:08:31 [debug] 7023#0: *35 [lua] balancer.lua:112: fetching targets for upstream: c81ff421-7a46-4d2d-9ce5-f560d6a4c455
2019/04/12 14:08:31 [debug] 7023#0: *35 [lua] balancer.lua:112: fetching targets for upstream: c81ff421-7a46-4d2d-9ce5-f560d6a4c455
2019/04/12 14:09:01 [crit] 7022#0: *7 [lua] balancer.lua:622: on_upstream_event(): failed creating balancer for buckets: timeout waiting for balancer for c81ff421-7a46-4d2d-9ce5-f560d6a4c455, context: ngx.timer
2019/04/12 14:09:01 [error] 7022#0: *7 [lua] balancer.lua:591: on_target_event(): target create: balancer not found for buckets, context: ngx.timer
2019/04/12 14:09:01 [error] 7022#0: *7 [lua] balancer.lua:591: on_target_event(): target create: balancer not found for buckets, context: ngx.timer
2019/04/12 14:09:01 [debug] 7022#0: *7 [lua] balancer.lua:409: create_balancer(): BALANCER_DBLESS_DEBUG - Marking creating with nil
2019/04/12 14:09:01 [debug] 7022#0: *7 [lua] balancer.lua:744: init(): BALANCER_DBLESS_DEBUG - after creating balancer
2019/04/12 14:09:01 [debug] 7022#0: *7 [lua] balancer.lua:752: init(): initialized 1 balancer(s), 0 error(s)
```

We can see that second worker (PID 7022) trying to create the balancer will get stuck for about 30s. All the requests going to that thread for the upstream will take a long to return, or timeout. With more workers (e.g., 4) the requests to that service will certainly timeout for a long time (minutes) until it finishes creating everything.

### Steps To Reproduce

1. use attached kong.yml as declarative config. The targets won't be accessible, but the problem can be reproduced by just starting kong with that config.
2. use more than 1 worker setting nginx_worker_processes on kong.conf
3. start kong and watch logs
4. also, could apply the attached patch to see the logs I mention above
5. with several workers, try to request:
```
http localhost:8000/v1/f174909b-40f1-4d71-ae5f-4fa86ff2b4ab.json
```
... it will certainly timeout (30s) the request in the first minutes after Kong started.

### Additional Details & Logs

- Kong version: 1.1.0
- Kong configuration: at the end of bug (attachments section)
- Operating system: Centos 7 (on vagrant VM)

##### 0001-Adding-logs-to-investigate-balancer-issue-with-dbles.patch 
```patch
From a602ebee3930b14b01fb01e37a37d824dafb5079 Mon Sep 17 00:00:00 2001
From: Daniel Liberman <daniel.liberman@dilettaaplicativos.com.br>
Date: Fri, 12 Apr 2019 11:39:52 -0300
Subject: [PATCH] Adding logs to investigate balancer issue with dbless mode

---
 kong/runloop/balancer.lua | 9 +++++++++
 kong/runloop/handler.lua  | 1 +
 2 files changed, 10 insertions(+)

diff --git a/kong/runloop/balancer.lua b/kong/runloop/balancer.lua
index 00ab5daf..14beb920 100644
--- a/kong/runloop/balancer.lua
+++ b/kong/runloop/balancer.lua
@@ -360,10 +360,12 @@ do
   -- @return The new balancer object, or nil+error
   create_balancer = function(upstream, recreate, history, start)
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 1")
     if balancers[upstream.id] and not recreate then
       return balancers[upstream.id]
     end
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 2")
     if creating[upstream.id] then
       local ok = wait(upstream.id)
       if not ok then
@@ -372,6 +374,7 @@ do
       return balancers[upstream.id]
     end
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 3")
     creating[upstream.id] = true
 
     local balancer, err = ring_balancer.new({
@@ -379,10 +382,12 @@ do
         dns = dns_client,
       })
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 4")
     if not balancer then
       return nil, err
     end
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 5")
     target_histories[balancer] = {}
 
     if not history then
@@ -393,6 +398,7 @@ do
       start = 1
     end
 
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - creating balancer - step 6")
     apply_history(balancer, history, start)
 
     create_healthchecker(balancer, upstream)
@@ -400,6 +406,7 @@ do
     -- only make the new balancer available for other requests after it
     -- is fully set up.
     balancers[upstream.id] = balancer
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - Marking creating with nil")
 
     creating[upstream.id] = nil
 
@@ -732,7 +739,9 @@ local function init()
   local oks, errs = 0, 0
   for name, id in pairs(upstreams) do
     local upstream = get_upstream_by_id(id)
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - before creating balancer")
     local ok, err = create_balancer(upstream)
+    log(DEBUG, "BALANCER_DBLESS_DEBUG - after creating balancer")
     if ok ~= nil then
       oks = oks + 1
     else
diff --git a/kong/runloop/handler.lua b/kong/runloop/handler.lua
index 9c9699f5..ba279da6 100644
--- a/kong/runloop/handler.lua
+++ b/kong/runloop/handler.lua
@@ -548,6 +548,7 @@ return {
 
 
       -- initialize balancers for active healthchecks
+      log(DEBUG, "BALANCER_DBLESS_DEBUG - calling balancer.init()")
       ngx.timer.at(0, function()
         balancer.init()
       end)
-- 
2.17.1
```

##### kong.yml
```yaml
_format_version: '1.1'

services:
- name: buckets
  url: http://buckets/
  routes:
  - name: root
    paths:
    - "/"

upstreams:
- name: buckets
  healthchecks:
    active:
      http_path: "/empty.json"
      unhealthy:
        interval: 5
      healthy:
        successes: 2
        interval: 0
        http_statuses:
        - 200
        - 301
        - 302
    passive: 
      healthy:
        successes: 2
        http_statuses:
        - 200
        - 301
        - 302
      unhealthy:
        tcp_failures: 2
        timeouts: 2
        http_failures: 2
        http_statuses:
        - 404
        - 429
        - 500
        - 501
        - 502
        - 503
  targets:
  - target: circuit-breakers-test2.s3-website.us-east-2.amazonaws.com:80
    weight: 1
  - target: circuit-breakers-test1.s3-website-sa-east-1.amazonaws.com:80
    weight: 1000
```
##### kong config
```json
{"plugins":{"enabled_in_cluster":[],"available_on_server":{"correlation-id":true,"pre-function":true,"request-transformer-advanced":true,"cors":true,"rate-limiting":true,"loggly":true,"ifood-kong-jwt-merchants-plugin":true,"hmac-auth":true,"zipkin":true,"bot-detection":true,"ifood-kong-remote-addr-plugin":true,"azure-functions":true,"request-transformer":true,"oauth2":true,"response-transformer":true,"ip-restriction":true,"statsd":true,"jwt":true,"basic-auth":true,"key-auth":true,"ifood-kong-prometheus-plugin":true,"ifood-kong-rbl-plugin":true,"http-log":true,"sample-cache-plugin":true,"generic-request-transformer":true,"ifood-kong-jwt-public-keys-plugin":true,"datadog":true,"tcp-log":true,"ifood-kong-jwt-plugin":true,"post-function":true,"prometheus":true,"acl":true,"kubernetes-sidecar-injector":true,"request-size-limiting":true,"response-ratelimiting":true,"udp-log":true,"file-log":true,"aws-lambda":true,"syslog":true,"ldap-auth":true,"request-termination":true}},"tagline":"Welcome to kong","configuration":{"plugins":["bundled","ifood-kong-jwt-merchants-plugin","ifood-kong-jwt-plugin","ifood-kong-rbl-plugin","ifood-kong-remote-addr-plugin","generic-request-transformer","request-transformer-advanced","sample-cache-plugin","ifood-kong-jwt-public-keys-plugin","ifood-kong-prometheus-plugin"],"client_ssl":false,"proxy_access_log":"\/vagrant\/kong-ce\/access.log","trusted_ips":["0.0.0.0\/0"],"prefix":"\/usr\/local\/kong","loaded_plugins":{"correlation-id":true,"pre-function":true,"request-transformer-advanced":true,"cors":true,"rate-limiting":true,"loggly":true,"ifood-kong-jwt-merchants-plugin":true,"hmac-auth":true,"zipkin":true,"bot-detection":true,"ifood-kong-remote-addr-plugin":true,"azure-functions":true,"request-transformer":true,"oauth2":true,"response-transformer":true,"ip-restriction":true,"statsd":true,"jwt":true,"basic-auth":true,"key-auth":true,"prometheus":true,"ldap-auth":true,"http-log":true,"sample-cache-plugin":true,"syslog":true,"ifood-kong-jwt-public-keys-plugin":true,"datadog":true,"ifood-kong-jwt-plugin":true,"ifood-kong-prometheus-plugin":true,"post-function":true,"file-log":true,"acl":true,"kubernetes-sidecar-injector":true,"request-size-limiting":true,"response-ratelimiting":true,"udp-log":true,"generic-request-transformer":true,"aws-lambda":true,"ifood-kong-rbl-plugin":true,"tcp-log":true,"request-termination":true},"cassandra_username":"kong","ssl_cert_key":"\/usr\/local\/kong\/ssl\/kong-default.key","dns_resolver":{},"pg_user":"kong","mem_cache_size":"20m","cassandra_data_centers":["dc1:2","dc2:3"],"nginx_admin_directives":{},"nginx_http_directives":[{"value":"prometheus_metrics 5m","name":"lua_shared_dict"}],"pg_host":"192.168.122.1","nginx_acc_logs":"\/usr\/local\/kong\/logs\/access.log","proxy_listen":["0.0.0.0:8000","0.0.0.0:8443 ssl"],"declarative_config":"\/etc\/kong\/kong.yml","client_ssl_cert_default":"\/usr\/local\/kong\/ssl\/kong-default.crt","ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/kong-default.key","db_update_frequency":5,"db_update_propagation":0,"stream_listen":["off"],"nginx_err_logs":"\/usr\/local\/kong\/logs\/error.log","cassandra_port":9042,"dns_order":["LAST","SRV","A","CNAME"],"dns_error_ttl":1,"headers":["server_tokens","latency_tokens"],"dns_valid_ttl":120,"cassandra_lb_policy":"RequestRoundRobin","nginx_optimizations":true,"origins":{},"database":"off","pg_database":"kong110_dbless","nginx_worker_processes":"2","lua_package_cpath":"","admin_acc_logs":"\/usr\/local\/kong\/logs\/admin_access.log","lua_package_path":".\/?.lua;.\/?\/init.lua;","nginx_pid":"\/usr\/local\/kong\/pids\/nginx.pid","upstream_keepalive":60,"cassandra_contact_points":["127.0.0.1"],"db_cache_ttl":0,"admin_access_log":"\/vagrant\/kong-ce\/admin_access.log","nginx_conf":"\/usr\/local\/kong\/nginx.conf","nginx_stream_directives":{},"proxy_listeners":[{"transparent":false,"ssl":false,"ip":"0.0.0.0","proxy_protocol":false,"port":8000,"http2":false,"listener":"0.0.0.0:8000"},{"transparent":false,"ssl":true,"ip":"0.0.0.0","proxy_protocol":false,"port":8443,"http2":false,"listener":"0.0.0.0:8443 ssl"}],"proxy_ssl_enabled":true,"cassandra_repl_factor":1,"dns_stale_ttl":4,"enabled_headers":{"latency_tokens":true,"X-Kong-Proxy-Latency":true,"Via":true,"server_tokens":true,"Server":true,"X-Kong-Upstream-Latency":true,"X-Kong-Upstream-Status":false},"admin_ssl_cert_default":"\/usr\/local\/kong\/ssl\/admin-kong-default.crt","ssl_ciphers":"ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256","db_resurrect_ttl":30,"pg_timeout":60000,"cassandra_consistency":"ONE","client_max_body_size":"0","admin_error_log":"\/vagrant\/kong-ce\/error.log","pg_ssl_verify":false,"dns_not_found_ttl":30,"pg_ssl":false,"admin_ssl_enabled":false,"error_default_type":"text\/plain","cassandra_repl_strategy":"SimpleStrategy","real_ip_header":"X-Real-IP","log_level":"debug","nginx_sproxy_directives":{},"ssl_cipher_suite":"modern","nginx_kong_conf":"\/usr\/local\/kong\/nginx-kong.conf","cassandra_schema_consensus_timeout":60000,"dns_hostsfile":"\/etc\/hosts","admin_listeners":[{"transparent":false,"ssl":false,"ip":"0.0.0.0","proxy_protocol":false,"port":8001,"http2":false,"listener":"0.0.0.0:8001"}],"nginx_proxy_directives":{},"ssl_cert":"\/usr\/local\/kong\/ssl\/kong-default.crt","cassandra_timeout":60000,"admin_ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/admin-kong-default.key","cassandra_ssl_verify":false,"dns_no_sync":false,"lua_ssl_verify_depth":1,"real_ip_recursive":"on","proxy_error_log":"\/vagrant\/kong-ce\/error.log","client_ssl_cert_key_default":"\/usr\/local\/kong\/ssl\/kong-default.key","nginx_daemon":"on","anonymous_reports":false,"kong_env":"\/usr\/local\/kong\/.kong_env","stream_listeners":{},"pg_port":5432,"nginx_kong_stream_conf":"\/usr\/local\/kong\/nginx-kong-stream.conf","client_body_buffer_size":"8k","ssl_preread_enabled":true,"lua_socket_pool_size":30,"ssl_cert_csr_default":"\/usr\/local\/kong\/ssl\/kong-default.csr","cassandra_keyspace":"kong","ssl_cert_default":"\/usr\/local\/kong\/ssl\/kong-default.crt","cassandra_ssl":false,"admin_listen":["0.0.0.0:8001"]},"version":"1.1.0","node_id":"4613561d-5c63-4ae3-93c0-396af421baf6","lua_version":"LuaJIT 2.1.0-beta3","prng_seeds":{"pid: 7023":285413248142,"pid: 7014":209769119018,"pid: 7022":222321021771},"timers":{"pending":2,"running":0},"hostname":"localhost.localdomain"}
```

**Comment by hishamhm** [comment]
@dliberman thanks for the detail report! I'll look into reproducing it.

**Comment by hishamhm** [comment]
Hi @dliberman, I have a PR with a proposed fix open here: https://github.com/Kong/kong/pull/4534

**Comment by dliberman** [comment]
@hishamhm, I just tested and it seems to be working fine with your changes. :)

After a little more investigation I came to another small bug. I found out I couldn't see any log messages when it had DB mode OFF. It was a small issue on load_declarative_config() ([here](https://github.com/Kong/kong/blob/bc48efeabc5d7460271d09d20e8e95b1b2e97a04/kong/init.lua#L254)) where it tried to load it regardless of database was set to off or not, if you passed the declarative file name on kong.conf. But I also see on master that you've fixed it already as well.

Thanks a lot for working on it!

CC: @UkiahSmith @carnei-ro
